### PR TITLE
[hardening] add option to enable Move VM paranoid checks

### PR DESCRIPTION
**Pull request by sblackshear**
These have a performance hit, but can catch transactions that hit bugs in the bytecode verifier.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -1071,6 +1071,8 @@ impl AuthorityState {
             adapter::new_move_vm(
                 epoch_store.native_functions().clone(),
                 epoch_store.protocol_config(),
+                self.expensive_safety_check_config
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:1074]: Are these checks expected to activate in dry runs?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs [hunk 2]
```diff
@@ -1943,9 +1943,14 @@ impl ExecutionComponents {
         metrics: Arc<ResolverMetrics>,
     ) -> Self {
         let native_functions = sui_framework::natives::all_natives();
+        let enable_move_vm_paranoid_checks = false;
```
Comment (lxfind) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:1946]: Why setting them to false if we already have the config?

File: crates/sui-core/src/authority.rs (hunk 1 above)
Comment (sblackshear) [reply crates/sui-core/src/authority.rs:1074]: Fair point--these should only ever activate for malicious code, no point in paying the cost for dry run.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs (hunk 2 above)
Comment (sblackshear) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1946]: I did not see how to access `ExpensiveSafetyCheckConfig` here (it is not inside `protocol_config`)

Comment (lxfind) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1946]: Right, you would need to pass it from the constructor of per epoch store

Comment (sblackshear) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1946]: I will do that.

File: crates/sui-config/src/node.rs [hunk 3]
```diff
@@ -287,9 +292,14 @@ impl ExpensiveSafetyCheckConfig {
         Self {
             enable_epoch_sui_conservation_check: true,
             force_disable_epoch_sui_conservation_check: false,
+            enable_move_vm_paranoid_checks: true,
         }
     }
 
+    pub fn force_enable_paranoid_checks(&mut self) {
```
Comment (lxfind) [inline crates/sui-config/src/node.rs:299]: No need for the `force_` prefix

**Review by lxfind — APPROVED** [review]

File: crates/sui-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap [hunk 4]
```diff
@@ -80,6 +80,7 @@ validator_configs:
       force-disable-epoch-sui-conservation-check: false
       enable-state-consistency-check: false
       force-disable-state-consistency-check: false
+      enable-move-vm-paranoid-checks: false
```
Comment (sblackshear) [inline crates/sui-config/tests/snapshots/snapshot_tests__network_config_snapshot_matches.snap:83]: @lxfind does this snapshot change mean that we need to version `ExpensiveSafetyCheckConfig`, or is this fine?

**Pull request merged**